### PR TITLE
feat(disk-fill): Adding the disk-fill experiment

### PR DESCRIPTION
**Pull request by ispeakc0de**
Signed-off-by: shubhamchaudhary <shubham.chaudhary@mayadata.io>

- Adding the disk-fill experiment

- fixes: litmuschaos/litmus#1693

### CODE REVIEW

**Comment by ispeakc0de** [comment]
**Logs**
```
INFO[2020-07-09T20:21:42+05:30] [PreReq]: Getting the ENV for the  experiment 
INFO[2020-07-09T20:21:42+05:30] [PreReq]: Updating the chaos result of disk-fill experiment (SOT) 
INFO[2020-07-09T20:21:42+05:30] The application informations are as follows   Label="run=nginx" Ramp Time=0 Namespace=shubham
INFO[2020-07-09T20:21:42+05:30] [Status]: Verify that the AUT (Application Under Test) is running (pre-chaos) 
INFO[2020-07-09T20:21:42+05:30] [Status]: Checking whether application pods are in running state 
INFO[2020-07-09T20:21:42+05:30] The running status of Pods are as follows     Status=Running Pod=nginx1-6876f9c9f8-lc7hn
INFO[2020-07-09T20:21:42+05:30] [Status]: Checking whether application containers are in running state 
INFO[2020-07-09T20:21:42+05:30] The running status of container are as follows  container=nginx Pod=nginx1-6876f9c9f8-lc7hn Status=Running
values 2147484   2147484 INFO[2020-07-09T20:21:42+05:30] [Info]: Details of application under chaos injection  ContainerName=nginx ephemeralStorageLimit=2147484 ephemeralStorageRequest=2147484 ContainerID=4d8aff28953f174dffe8502afe123eb4342c07ffccea88a88004e5e403f56185 PodName=nginx1-6876f9c9f8-lc7hn NodeName=minikube
INFO[2020-07-09T20:21:42+05:30] [Status]: Checking the status of the helper pod 
INFO[2020-07-09T20:21:42+05:30] [Status]: Checking whether application pods are in running state 
INFO[2020-07-09T20:21:50+05:30] The running status of Pods are as follows     Pod=disk-fill-aamuou Status=Running
INFO[2020-07-09T20:21:50+05:30] [Status]: Checking whether application containers are in running state 
INFO[2020-07-09T20:21:50+05:30] The running status of container are as follows  container=disk-fill Pod=disk-fill-aamuou Status=Running
INFO[2020-07-09T20:21:50+05:30] used ephemeral storage space: 28             
INFO[2020-07-09T20:21:50+05:30] ephemeral storage size to be filled: 1717959 
429489+0 records in
429489+0 records out
1759186944 bytes (1.8 GB, 1.6 GiB) copied, 13.3838 s, 131 MB/s
INFO[2020-07-09T20:22:04+05:30] [Wait]: Waiting for the 60s after injecting chaos 
INFO[2020-07-09T20:23:04+05:30] [Cleanup]: Deleting the helper pod           
INFO[2020-07-09T20:23:36+05:30] [Confirmation]: The application pod has been deleted successfully 
INFO[2020-07-09T20:23:36+05:30] [Status]: Verify that the AUT (Application Under Test) is running (post-chaos) 
INFO[2020-07-09T20:23:36+05:30] [Status]: Checking whether application pods are in running state 
INFO[2020-07-09T20:23:36+05:30] The running status of Pods are as follows     Pod=nginx1-6876f9c9f8-lc7hn Status=Running
INFO[2020-07-09T20:23:36+05:30] [Status]: Checking whether application containers are in running state 
INFO[2020-07-09T20:23:36+05:30] The running status of container are as follows  Pod=nginx1-6876f9c9f8-lc7hn Status=Running container=nginx
INFO[2020-07-09T20:23:36+05:30] [The End]: Updating the chaos result of disk-fill experiment (EOT)
```

File: chaoslib/litmus/disk_fill/disk-fill.go [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+
+	// it will return the ephemeral storage size for the target container
+	command := fmt.Sprintf("du /diskfill/" + containerID)
+
+	req := clients.KubeClient.CoreV1().RESTClient().Post().
+		Resource("pods").
+		Name("disk-fill-" + experimentsDetails.RunID).
+		Namespace(experimentsDetails.ChaosNamespace).
+		SubResource("exec")
+	scheme := runtime.NewScheme()
```
Comment (ksatchit) [inline chaoslib/litmus/disk_fill/disk-fill.go:41]: Can we describe this step please

File: chaoslib/litmus/disk_fill/disk-fill.go [hunk 2]
```diff
@@ -0,0 +1,497 @@
...
+	// deriving the ephemeral storage size to be filled
+	sizeTobeFilled := GetSizeToBeFilled(experimentsDetails, usedEphemeralStorageSize, int(ephemeralStorageLimit))
+
+	log.Infof("ephemeral storage size to be filled: %v", strconv.Itoa(sizeTobeFilled))
+
+	if sizeTobeFilled > 0 {
+		// Creating files to fill those size
+		FileCreation(experimentsDetails, clients, containerID, sizeTobeFilled)
+	} else {
+		log.Warn("jagha nahi hai, housefull")
```
Comment (ksatchit) [inline chaoslib/litmus/disk_fill/disk-fill.go:169]: its cool! may an english equivalent?

File: chaoslib/litmus/disk_fill/disk-fill.go [hunk 3]
```diff
@@ -0,0 +1,497 @@
...
+	req.VersionedParams(&apiv1.PodExecOptions{
+		Command:   strings.Fields(command),
+		Container: "disk-fill",
+		Stdin:     false,
+		Stdout:    true,
+		Stderr:    true,
+		TTY:       false,
+	}, parameterCodec)
+
+	exec, err := remotecommand.NewSPDYExecutor(clients.KubeConfig, "POST", req.URL())
```
Comment (ksatchit) [inline chaoslib/litmus/disk_fill/disk-fill.go:444]: Just a wishlist - would love to see a exec.go that probably parameterizes all this. So I can just do Exec.{{command}}.

File: chaoslib/litmus/disk_fill/disk-fill.go [hunk 4]
```diff
@@ -0,0 +1,497 @@
...
+
+	err = exec.Stream(remotecommand.StreamOptions{
+		Stdin:  nil,
+		Stdout: stdout,
+		Stderr: stderr,
+		Tty:    false,
+	})
+
+	if err != nil {
+		errorCode := strings.Contains(err.Error(), "143")
```
Comment (ksatchit) [inline chaoslib/litmus/disk_fill/disk-fill.go:461]: We may want to generalize this considering the disk-fill itself might become more generic (i.e., not just ephemeral. Also fill mode can be something other than dd)

**Pull request merged**